File: bench/python/ops/bench_nonmaximumsuppression.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA NonMaximumSuppression operator benchmark - Python equivalent of BenchNonMaximumSuppression.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cupy as cp  # noqa: E402
import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    parse_shape,
    create_tensor,
    create_stream_cache,
    run_benchmark,
    _lcg_fill,
    _LCG_SEED,
)


def nms(state):
    """NonMaximumSuppression operator benchmark matching C++ BenchNonMaximumSuppression.cpp"""

    shape = parse_shape(state.get_string("shape"))
    device_id = state.get_device()
    score_threshold = state.get_float64("scoreThreshold")
    iou_threshold = state.get_float64("iouThreshold")

    N, num_boxes = shape

    state.add_global_memory_reads(N * num_boxes * (8 + 4))
    state.add_global_memory_writes(N * num_boxes * 1 * 2)

    get_stream = create_stream_cache()

    src_bb = cvcuda.Tensor((N, num_boxes), cvcuda.Type._4S16, "NB")
    with cp.cuda.Device(device_id):
        # Deterministic per-element [10, 50] cycle mirroring the C++
        # BenchNonMaximumSuppression lambda exactly (n=batch, b=box, c=channel
        # → 10 + (b*7 + c*11 + n*3) % 41).
        n_idx, b_idx, c_idx = cp.indices((N, num_boxes, 4), dtype=cp.int64)
        bb_data = (10 + (b_idx * 7 + c_idx * 11 + n_idx * 3) % 41).astype(cp.int16)
        cp.copyto(
            cp.asarray(src_bb.cuda()).view(cp.int16).reshape(N, num_boxes, 4), bb_data
        )

    with cp.cuda.Device(device_id):
        # srcSc: deterministic LCG over float [-1, +1] (bit-identical to C++
        # RandomValues<float>() which routes through the GPU LCG fast path).
        sc_data = cp.empty((N, num_boxes), dtype=cp.float32)
        _lcg_fill(sc_data, _LCG_SEED)
        src_sc = cvcuda.as_tensor(sc_data, "NB")

    dst_mk = create_tensor(
        (N, num_boxes), cvcuda.Type.U8, device_id, layout="NB", fill_mode=0
    )

    def run(launch):
        cvcuda.nms_into(
            dst_mk,
            src_bb,
            src_sc,
            score_threshold,
            iou_threshold,
            stream=get_stream(launch),
        )

    return run


if __name__ == "__main__":
    run_benchmark("nonmaximumsuppression", nms)

File: bench/cpp/ops/BenchNonMaximumSuppression.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchNonMaximumSuppressionConfig.hpp"

#include <cvcuda/OpNonMaximumSuppression.hpp>

#include <nvbench/nvbench.cuh>

template<typename T, typename S = float, typename M = uint8_t>
inline void nms(nvbench::state &state, nvbench::type_list<T>)
try
{
    int2 shape  = benchutils::GetShape<2, int2>(state.get_string("shape"));
    auto scThr  = static_cast<float>(state.get_float64("scoreThreshold"));
    auto iouThr = static_cast<float>(state.get_float64("iouThreshold"));

    // R/W bandwidth rationale:
    // 1 read of scores (F32) to mask out lower scores boxes + 1 read of boxes (4S16) for IoU threshold
    // 2 writes of masks (U8) by score and IoU thresholds
    state.add_global_memory_reads(shape.x * shape.y * (sizeof(T) + sizeof(S)));
    state.add_global_memory_writes(shape.x * shape.y * sizeof(M) * 2);

    cvcuda::NonMaximumSuppression op;

    // clang-format off

    // Config only specifies the Tensor variant (inputKind=Tensor); no ImageBatchVarShape support
    nvcv::Tensor srcBB({{shape.x, shape.y}, "NB"}, benchutils::GetDataType<T>());
    nvcv::Tensor srcSc({{shape.x, shape.y}, "NB"}, benchutils::GetDataType<S>());
    nvcv::Tensor dstMk({{shape.x, shape.y}, "NB"}, benchutils::GetDataType<M>());

    // srcBB: deterministic per-element [10, 50] cycle. The Python bench
    // mirrors this with `10 + (b*7 + c*11 + n*3) % 41` over (N, num_boxes, 4),
    // giving bit-identical bytes for the box coordinates.
    benchutils::FillTensor<T>(srcBB, [](const long4_16a &c) {
        long n = c.x;
        long b = c.y;
        auto v = [&b, &n](long ch) { return (short)(10 + (b * 7 + ch * 11 + n * 3) % 41); };
        return T{v(0), v(1), v(2), v(3)};
    });
    benchutils::FillTensor<S>(srcSc, benchutils::LcgValues<S>());
    // Output tensor - zero-fill (output only, no need for random values)
    benchutils::FillTensor<M>(dstMk, [](const long4_16a &) { return M{0}; });

    benchutils::warmup_and_exec(state, BENCH_NONMAXIMUMSUPPRESSION_WARMUP_ITERATIONS,
        [&op, &srcBB, &dstMk, &srcSc, &scThr, &iouThr](cudaStream_t s) { op(s, srcBB, dstMk, srcSc, scThr, iouThr); });
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(nms, NVBENCH_TYPE_AXES(BENCH_NONMAXIMUMSUPPRESSION_TYPES))
BENCH_NONMAXIMUMSUPPRESSION_AXES;
